### ISSUE TITLE
path.parse is not a function in browser environment

### ISSUE THREAD

**Issue by sachinkumar121** [seald/nedb#50]
**Describe the bug**
I am using this package in my vue.js application and it always gives the error of "path.parse is not a function". 

**To Reproduce**
- Install the latest version of the package and run the following in one of the modules.
```
const db = new Datastore({ 
    filename: "any_random_string",
    // autoload: true 
});
await db.loadDatabaseAsync(); // this line making the issue or if I comment out autoload option. 
console.log(db)
```

**Environment info**
- Runtime: Chrome Browser
- Version: 120.0.
- OS: macOS 12.7.1

**Comment by arantes555** [comment]
Hello @sachinkumar121 .

How do you import the module ? Do you use a bundler or no ?

- If you don't use a bundler, and use the module directly, I guess you just used the `index.js`. This won't work. For use in a browser, you must use the browser bundle : https://github.com/seald/nedb#browser-bundle
- If you do use a bundler, it should use the browser bundle automatically. If that's not the case, do let us know.

**Comment by tex0l** [comment]
Hi,
I think this is related to this error: https://github.com/seald/nedb/issues/40

The problem is simple, the native module `path` is used internally, also other native modules are used (`events`, `util`), see:
- https://github.com/seald/nedb/blob/77d1b141806ae0acdb96afcac118f253d34f4997/lib/datastore.js#L1
- https://github.com/seald/nedb/blob/77d1b141806ae0acdb96afcac118f253d34f4997/lib/datastore.js#L2
- https://github.com/seald/nedb/blob/77d1b141806ae0acdb96afcac118f253d34f4997/lib/cursor.js#L2
- https://github.com/seald/nedb/blob/77d1b141806ae0acdb96afcac118f253d34f4997/lib/persistence.js#L1
- https://github.com/seald/nedb/blob/77d1b141806ae0acdb96afcac118f253d34f4997/lib/persistence.js#L2

And probably some global objects that may not exist in a browser (like Buffer, setImmediate, process, etc.)

Therefore, if you don't use the prebuilt already polyfilled browser version, you should polyfill it yourself.

If you use webpack, you can check out the [webpack configuration](https://github.com/seald/nedb/blob/master/webpack.config.js) used to build the prebuilt browser version provided in the npm package.

A solution could be to replace these uses of native node modules by helpers in browser-compatible JS, as it is done for other utils: https://github.com/seald/nedb/blob/master/browser-version/lib/customUtils.js

**Comment by sachinkumar121** [comment]
HI @arantes555 I am testing this package on a vue.js application using VITE[https://stackblitz.com/edit/vitejs-vite-9cnj9c]. 
In the sample component I have added the following and it throws the error.

The localforage should select the best storage method based on my browser but it seems like it is trying to parse the non-reachable dir.

```
<template>
  <div>
    <h1>Hello world!</h1>
  </div>
</template>

<script>
import Datastore from '@seald-io/nedb';

export default {
  async mounted() {
    const db = new Datastore({ 
      autoload: true,
      filename: "random_string_based_on_user" 
    });
  }
}
</script>
```

vite.config.js

```
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [vue()],
})
```

**Comment by tex0l** [comment]
Hi!

Indeed, importing like this with no additional vite config won't work.

I'm currently working on a patch to solve the 'path' problem, you are right that it should not be used in a browser env, and it will be so in a future version, but the native modules `util` and `events` will still be used anyway.

You have three choices:
- use a pre-polyfilled bundled by importing `@seald-io/nedb/browser-verison/out/nedb.js` which puts `Datastore` as a global variable.
- import the non-bundled version and polyfill it yourself by modifying your vite config with https://www.npmjs.com/package/vite-plugin-node-polyfills
- add the `path`, `util` and `events` browserify packages to your dependencies and it _should_ work since their use within nedb is very simple (not tested).

**Comment by sachinkumar121** [comment]
If it is related to manually adjustment of pollyfill then I guess I have the solution.

```
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'
import { NodeGlobalsPolyfillPlugin } from '@esbuild-plugins/node-globals-polyfill'


// https://vitejs.dev/config/
export default defineConfig({
  plugins: [vue()],
  optimizeDeps: {
    esbuildOptions: {
      plugins: [
        NodeGlobalsPolyfillPlugin({
          process: true,
          path: true,
          buffer: true
        })
      ]
    }
  },
})
```

**Comment by tex0l** [comment]
Something like this should work yes, you should include `events`, `process` and `path` in the polyfills, nothing else. If not, it may indicate that the files declared in the browser field of the package.json are not used.

**Comment by sachinkumar121** [comment]
> @sachinkumar121 After a quick check, could you please try with 4.0.2 instead of 4.0.3 ? It's not impossible that 4.0.3 introduced something that breaks in the browser.

FYI, @arantes555  we have tested the package's latest version(v4.0.3) and it is not compatible with polyfill but working fine with 4.0.2.

**Comment by tex0l** [comment]
Hi!
I managed to reproduce the issue using the vite config you provided.

However, when using the plugin I mentioned (`'vite-plugin-node-polyfills'`) instead, it works even in v4.0.3:

```
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'
import { nodePolyfills } from 'vite-plugin-node-polyfills'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    nodePolyfills(),
    vue(),
  ]
})
```

**Comment by tex0l** [comment]
I released 4.0.4 which fixes the regression of 4.0.3.

**Comment by sachinkumar121** [comment]
Great @tex0l
